### DISCUSSION TITLE
Instrument management restructuring

### DISCUSSION THREAD

**Discussion by msmttchr** [pymeasure/pymeasure#617] [Ideas]
I would like to start a discussion to brainstorm and converge on possible evolution of `Instrument` and `Adapter` classes.

# Background
There are several Issues and related discussion that touch various aspects of the instruments management in `PyMeasure`.
Most notably, we have the following issues:
- #512 
- #219 
- https://github.com/pymeasure/pymeasure/pull/430#issuecomment-824973185
- #567 
- #416 
- https://github.com/pymeasure/pymeasure/issues/416#issuecomment-1025090693
- https://github.com/pymeasure/pymeasure/pull/606#discussion_r841061884
- #587 
- #506 

I think these are providing enough food for extensive discussion, I would start apologizing for the long thread where I try to explain my proposal.

## `Instrument` class
`Instrument` class will support few interface methods which we call _Basic_ for brevity.
_Basic_ methods are not implemented by default and for the sake of discussion they could be `reset`, `check_errors`, `shutdown`,..

In addition, it defines interface towards physical layer via `read`/`write` (ASCII/binary)/`ask`/`values` methods.
In particular, `ask` and `values` methods are implemented at `Instrument` level and not at `Adapter` level.
These methods can be overridden in sub-classes.
In particular, the `ask` method should deal with thread safety since different thread can take over between `write` and `read` and it will not be based on `query` method inside PyVISA.

The access to physical layer interface is provided, as it is today, by an `Adapter` instance generated at runtime or passed as parameter to capture the fact that interface is something outside the instrument and can be swapped.

## SCPI instrument

These instruments ared managed by a dedicated class and the parameter `includeSCPI` is removed from the `Instrument` interface.
- `SCPIBase` class: derived from `Instrument` class and it implements a subset of mandatory SCPI commands and _Basic_ methods

All SCPI instruments are subclass of `SCPIBase`.

A runtime check for SCPI compliance could be:
`isinstance(inst, SCPIBase)`

## Non SCPI instrument

- Implement Basic interface
- Implement instrument specific interface

Non SCPI instruments are subclass of `Instrument`.

## Adapters

Adapters are hardware interface used to communicate from PC to instrument.
Some interfaces are native in the PC, e.g. LAN, USB other requires some external hardware, e.g. USB to RS232, or USB to GPIB.

Normally, `PyVISA` is used to manage the physical interface with some exceptions being `Prologix` adapters (ethernet and USB) or `Arduino` GPIB.

GPIB is very old interface being replaced by USB/LAN/PCIe, unfortunately it is still used and popular as such dropping support for cheap GPIB could be an issue.

In such case, we can provide:
- predefined custom adapter for `Prologix`
- predefined custom adapter for `Arduino`
- template for other unknown adapters

[Eventually `PyVISA` could add support for `Prologix`]

My proposal would be to keep current `Adapter` classes with basic support for message based communication, like read/write, read binary/write binary and moving other methods like ask, values to `Instrument` class.

Keep and update `VISAAdapter`, `PrologixAdapter` and possibly `SerialAdapter` for very simple implementation without VISA.

We need to pay attention to parameters like timeout that currently have ms unit in `PyVISA` GPIB and seconds in `PyVISA` Serial.

We need also to offer some extra flexibility like adding an optional delay after commands and logging instrument messages.

We could define a new address for `Prologix`, e.g. something like `"PRX29::27::INSTR"`.

Non message-based instrument can be left out, but probably an implementation as new `Adapter` could be possible if the need arises.

Instrument developers should not need to develop custom adapters.

## Instrument channels

They implement part of the instrument functions that is replicated _n_ times in the instrument.
E.g.: Multi channel power supply

I would assume that a new `InstrumentChannel` derived from `Instrument` should be enough and if needed some customization can be applied to:
- Add a prefix,suffix,infix to all write/read commands.
Dynamic properties should be supported since the parent class supports it.
Typically the channel implementation override `write` method.

## Instrument families
[Here I use the term families, but really I mean classes, but I avoid the term classes would create clash with the same python term]
These are instruments providing similar features, eg. power supply, multimeter, spectrum analyser, etc.
They can be SCPI compliant or not SCPI compliant.
Even instrument that are SCPI compliant, use different interpretation of the spec and as a result, it's difficult to consider SCPI as a  solid API.
The strategy should be to define an interface abstract enough and as simple as possible, avoiding implementing special features offered by a certain instrument only or features that can be implemented in software. Here the rationale is to exploit as much as possible hardware part and implement all the remaining software part in python to make interoperability easier.

An instrument can derive from one or more (composite instruments) instrument families.
In the instrument implementation, features specific to that instrument can be provided in addition to the one defined by the family.

Final though, here I would favor composition over inheritance, but I have to say that I am not fully able to propose something regarding the granularity. I have seen approaches like yaq (https://github.com/pymeasure/pymeasure/issues/416#issuecomment-803383314) that are at very fine grain level, while I would prefer some coarse grain approach.

## Migration path
Assuming the proposal is accepted, presumably after several reviews, a step by step approach is recommended.
My proposal would be to:

1. Modify Instrument class, Adapter class and implement SCPI base
2. Update the existing instrument code base
3. Update user level documentation
4. Implement Instrument channels support
5. Update the existing instrument code base
6. Update user level documentation
7. Implement some Instrument families support
5. Update the existing instrument code base
6. Update user level documentation

I would say that step 1-3 can be phase 1, step 4-6 can be phase 2 and step 7-9 can be phase 3.
They could also run in parallel, but it depends on the available bandwidth.

**Comment by bilderbuchi** [comment]
Good initiative, I mostly agree. I'll remark on some points in separate comments, maybe the Discussions interface lets us keep the threads separate, let's see. :-)

**Reply by bilderbuchi** [reply]
Let's see how it goes with this discussion format. Maybe we'll need to hop on a call at some point. Or, you could turn this discussion after 1 iteration (or whenever the replies and re-replies explode :D) in here into a committed design document, that we could update/discuss in a PR until satisfied (similar to [NEPs](https://numpy.org/neps/nep-0000.html) and equivalent), and then start implementation.

**Reply by bilderbuchi** [reply]
I whipped together this png, which should be editable on https://www.diagrams.net/ - download it and "open from"-"device" with the web application. Then edit to your heart's content.
(I did not yet fill in attributes, methods, properties because it's late now :D)
It's also possible to store on Github, but I did not try, I don't know what happens if someone else then would edit it?

![NewHierarchy drawio](https://user-images.githubusercontent.com/327442/165624135-614a8bd5-3a08-4cee-b1c0-1c469ecc8882.png)
.

**Reply by msmttchr** [reply]
> Good initiative, I mostly agree. I'll remark on some points in separate comments, maybe the Discussions interface lets us keep the threads separate, let's see. :-)

@bilderbuchi, just to let you know that I noticed your feedback (thank you), but I am a bit busy in this period, so I would need quite some time to answer :(

**Reply by bilderbuchi** [reply]
Don't worry about that! Same with me as you know. Thanks for the notice, though!

**Comment by bilderbuchi** [comment]
## Instrument class

> Instrument class will support few interface methods which we call Basic for brevity.
> Basic methods are not implemented by default and for the sake of discussion they could be reset, check_errors, shutdown,..

I'm not sure why the Basic methods should not be defined in Instrument? Would then any non-SCPI instrument need to reimplement those? I think later we should get a concrete list together, right now I'm not 100% clear on the intent behind this.

> In particular, ask and values methods are implemented at Instrument level and not at Adapter level.

Agreed, that makes sense and would be an improvement, and also agreed to get rid of the use of the PyVISA query method, this inconsistency has just been annoying.

> In particular, the ask method should deal with thread safety since different thread can take over between write and read and it will not be based on query method inside PyVISA.

Disagree, or rather a refinement of the point. Thread safety (let's call this a lock because that's the mechanism) must be handled on the level of `read`/`write` calls with one lock to avoid two coincident `write` _or_ `read` calls colliding (something which locks on ask() won't achieve). 
There can/should probably be _another_ kind of lock on `ask` to make sure that its `read` can follow the associated `write` without another thread's call interfering. This is what you meant I think. This will need some thought to avoid race conditions and deadlocks and probably some nested locks.

**Comment by bilderbuchi** [comment]
## SCPI/NonSCPI

> SCPIBase class: derived from Instrument class and it implements a subset of mandatory SCPI commands and Basic methods

Propose to call this `SCPIInstrument` to mirror `Instrument`.
I'm wondering why you prefer to have an additional optional layer in the class hierarchy, instead of the approach with `includeSCPI` flag which we have now and which works well afaik. What's the advantage we gain?

> Non SCPI instrument
>    Implement Basic interface
>    Implement instrument specific interface
> Non SCPI instruments are subclass of Instrument.

I notice we are implementing the "Basic" interface twice -- why?

**Comment by bilderbuchi** [comment]
## Adapters

> Normally, PyVISA is used to manage the physical interface with some exceptions being Prologix adapters (ethernet and USB) or Arduino GPIB.

Agreed, PyVISA as first choice makes sense.

> GPIB is very old interface being replaced by USB/LAN/PCIe, unfortunately it is still used and popular as such dropping support for cheap GPIB could be an issue.

Agreed, we should definitely _not_ drop GPIB support.

> In such case, we can provide:

>    predefined custom adapter for Prologix
>    predefined custom adapter for Arduino
>    template for other unknown adapters

I think we should consider keeping `Serial` around for situations where a VISA installation is not an option.

> [Eventually PyVISA could add support for Prologix]

Agreed, I think we should see if we find interested folks to pick up https://github.com/pyvisa/pyvisa/issues/112 -- maybe @paulgoulain ?

> My proposal would be to keep current Adapter classes with basic support for message based communication, like read/write, read binary/write binary and moving other methods like ask, values to Instrument class.

:+1:

> Keep and update VISAAdapter, PrologixAdapter and possibly SerialAdapter for very simple implementation without VISA.

:+1: the others should be _deprecated_ with a proper warning message (`FutureWarning`) and upgrade path.

> We need also to offer some extra flexibility like adding an optional delay after commands and logging instrument messages.

yeah, having concrete read/write methods in `Instrument` that _delegate_ to `Adapter` methods should allow us to do some nice things centrally. Will have to check/adapt overrides already in the codebase, though.

> We could define a new address for Prologix, e.g. something like "PRX29::27::INSTR".

Let's try https://github.com/pyvisa/pyvisa/issues/112  first, before we come up with a custom protocol/addressing scheme.

> Non message-based instrument can be left out, but probably an implementation as new Adapter could be possible if the need arises.

If we can, I'd like to keep frame-based comms in mind for easy incorporation later (cf. #239 which has been on my wish-todo-list for ages). Not sure how/if possible at this point, though.

> Instrument developers should not need to develop custom adapters.

You propose that any customizations (I think "protocol quirks" is the established name by now) end up in the instrument class, right?

**Comment by bilderbuchi** [comment]
## Instrument channels

> I would assume that a new InstrumentChannel derived from Instrument should be enough and 

Actually, it should not derive from Instrument (or SCPIInstrument) as it would then inherit all the stuff (adapter, basic methods) that are not relevant for a Channel. A channel is most often (always?) a set of properties plus a prefix/suffix/override for the commands exchanged.
I think a `ChannelBase` root class is a better idea -- we seem to be doing fine without so far with just an Instrument `parent` ctor argument, but we need to sort out dynamic overrides which could be separate in another base class both for `ChannelBase` and `Instrument`. I think some documentation and a light `ChannelBase` to inherit from would be sufficient.

**Reply by bilderbuchi** [reply]
I think the main question re: channels will be if we want to standardize access to channels, and if yes, how.
* `myinstr.ch<N>.prop`
* `myinstr.channels[N].prop`
* Do all Instrument implicitly have a channel (I think no because it's too complicated), and if yes how does access work if only one channel is avaible, i.e. does the above then become `myinstr.prop`?

**Comment by bilderbuchi** [comment]
## Instrument families

> [Here I use the term families, but really I mean classes, but I avoid the term classes would create clash with the same python term]

I like that term! :-)

> These are instruments providing similar features, eg. power supply, multimeter, spectrum analyser, etc.
They can be SCPI compliant or not SCPI compliant.
Even instrument that are SCPI compliant, use different interpretation of the spec and as a result, it's difficult to consider SCPI as a solid API.

I think we can handle that by deleting non-supported methods in affected instruments, Python is so delightfully dynamic :D

> The strategy should be to define an interface abstract enough and as simple as possible, avoiding implementing special features offered by a certain instrument only or features that can be implemented in software. 

Agreed, I think the main work will be to come up with
* a set of families
* what the standard names should be for each (also also across families!), and their behaviour (would be a good point to harmonize naming throughout) `out` vs. `output` vs `output_enabled` vs `on`,...

> Here the rationale is to exploit as much as possible hardware part and implement all the remaining software part in python to make interoperability easier.

I'm afraid I don't get that point? :confused:

> An instrument can derive from one or more (composite instruments) instrument families.

:+1: (I guess you're thinking of SMUs?), probably best implemented as Mixin classes, although the discussion in #567 is highly relevant here.

> In the instrument implementation, features specific to that instrument can be provided in addition to the one defined by the family.

of course, agreed

> Final though, here I would favor composition over inheritance, but I have to say that I am not fully able to propose something regarding the granularity. I have seen approaches like yaq (https://github.com/pymeasure/pymeasure/issues/416#issuecomment-803383314) that are at very fine grain level, while I would prefer some coarse grain approach.

I'm not sure how we would do composition here considering we use properties so much, which are actually a class-level construct. I'm open to ideas, though.

**Comment by bilderbuchi** [comment]
## Migration path

Agreed on the phases etc. We should definitely keep a handle on the scope and ensure we can do it well in phases, I suspect this will be a rather large effort.

**Comment by BenediktBurger** [comment]
One additional thought to consider while restructuring:
Quite a few devices respond to every message written. For those devices, you have to add `check_set_errors=True` for every single `Instrument.control` or `Instrument.setting`, and you have to make `check_errors()` reading the rest of the message. It's easy to forget to add this check while writing or modifying such an instrument.
I propose to have some central control, whether a property write (of control or setting) triggers an automatic read or not. It can be a variable (i.e. `self.check_set_errors`) or changing `Instrument.control` for these Instruments such, that it automatically calls a  `self.read_response()` method.
How we implement it, depends on the decided upon path, be it mixin, subclassing etc. I want to raise this issue to take into consideration as well.

**Reply by bilderbuchi** [reply]
> Quite a few devices respond to every message written. For those devices, you have to add `check_set_errors=True` for every single `Instrument.control` or `Instrument.setting`, and you have to make `check_errors()` reading the rest of the message. It's easy to forget to add this check while writing or modifying such an instrument.

Either my IDE's code search is broken, or I want to say "citation needed" -- I'm seeing `check_set_errors` only about 11 times in instrument implementations, primarily in one agilent instrument?
If you forget to add this check, you'll notice quickly because the instrument will not work as expected, right? Because there's this response left in the read buffer.

**Reply by BenediktBurger** [reply]
The github search is broken... Also TC038 requires check_set_errors.
Another instrument wrote its own adapter, which reads the response in the adapter.write, which causes more problems, but the author did not know better, I guess.
I referred to devices I use in the lab (which are not yet in pymeasure). Around half of them respond always.

Yes, if you write tests, you should notice. But that way, you cannot forget.

**Reply by bilderbuchi** [reply]
No, this was PyCharm. Anyway, I double-checked in the package directory:
```
grep -rIn check_set_errors .

./instruments/hp/hp8116a.py:341:        check_set_errors=True
./instruments/rohdeschwarz/sfm.py:248:        check_set_errors=True,
./instruments/hcp/tc038.py:152:        check_set_errors=True,
./instruments/agilent/agilent4156.py:165:        check_set_errors=True,
./instruments/agilent/agilent4156.py:182:        check_set_errors=True,
./instruments/agilent/agilent4156.py:197:        check_set_errors=True,
./instruments/agilent/agilent4156.py:212:        check_set_errors=True,
./instruments/agilent/agilent4156.py:930:        check_set_errors=True,
./instruments/agilent/agilent4156.py:961:        check_set_errors=True,
./instruments/fakes.py:51:                check_set_errors=False, check_get_errors=False,
./instruments/fakes.py:78:                                  check_set_errors=check_set_errors,
./instruments/instrument.py:142:                         'check_set_errors')
./instruments/instrument.py:287:                check_set_errors=False,
./instruments/instrument.py:312:        :param check_set_errors: Toggles checking errors after setting
./instruments/instrument.py:395:                 check_set_errors=check_set_errors,
./instruments/instrument.py:414:            if check_set_errors:
./instruments/instrument.py:467:                check_set_errors=False, dynamic=False,
./instruments/instrument.py:483:        :param check_set_errors: Toggles checking errors after setting
./instruments/instrument.py:495:                                  check_set_errors=check_set_errors,
./instruments/keithley/keithley2700.py:110:        check_set_errors=True,
./instruments/keithley/keithley2700.py:124:        check_set_errors=True
./adapters/protocol.py:89:                "Maybe a property's 'check_set_errors' is not accounted for, "
``` 
22 hits, 11 of which are in the "core" part (+ 1 in a test, not shown). The one hit in tc038 is included.

> Another instrument wrote its own adapter, which reads the response in the adapter.write, which causes more problems, but the author did not know better, I guess.
I referred to devices I use in the lab (which are not yet in pymeasure). Around half of them respond always.

Thanks for the clarification!

Is that a hard "always"? I.e., these instruments all respond to _every_ write (not necessarily through a control/...) with a confirmation message? (I mean, your post implies a "no", I just want to make sure I understand)

Otherwise, it could be challenging to disambiguate, as you can write properties also in the conventional Python way with decorators. If they only respond to property-related writes, but not other commands (e.g. in methods), that seems quite puzzling.

It's good that you raise that concern! Maybe we could handle it similarly to `Adapter.preprocess_reply`, which we introduced for a similar reason -- see the discussion in #331.

**Reply by BenediktBurger** [reply]
> Is that a hard "always"? I.e., these instruments all respond to _every_ write (not necessarily through a control/...) with a confirmation message? (I mean, your post implies a "no", I just want to make sure I understand)

Yes these devices respond **always** and therefore, yes, not only controls, but also python properties setter need to read afterwards. Thanks for raising this issue.

**Reply by bilderbuchi** [reply]
Then I would handle it equivalently to `Adapter.preprocess_reply` (I guess soon `Instrument.preprocess_reply`) -- introduce (after the adapter rework) an `Instrument.check_set_errors` boolean, default `None`, that gets applied if set, and can be overridden by a `check_set_errors` defined in individual controls/.... (to allow for the situation where not all writes trigger a response)
It will be a bit tricky because the property factories have a `check_set_errors=False` default -- how to disambiguate between excluding _some_ property from the set-check, and the user just not having used that arg?

**Reply by BenediktBurger** [reply]
(Ab)Using check_errors (designed to **ask** for errors) to read the response does not seem right to me.
Maybe we could change the setter (for those instruments) to call `ask` instead of `write`? That seems to be more natural (obviously the read/ask method should check whether the response is an acknowledgement or error)

**Reply by bilderbuchi** [reply]
Hm, I see what you mean. Maybe have something like `check_set_ack`, which consumes/checks acknowledgement messages on every set?

Question: If you sent a query that expects a data response (like e.g. a voltage value), do you _also_ get a separate response/acknowledgement on the write, before/after getting the data message?

**Reply by BenediktBurger** [reply]
> Question: If you sent a query that expects a data response (like e.g. a voltage value), do you _also_ get a separate response/acknowledgement on the write, before/after getting the data message?

No. You always get exactly one response which contains the acknowledgment/error and (if you asked for a value and it succeeded) the value.
One device, for example, echoes the command back, if successful, otherwise "ERR". This is followed by ":" and the return value(if any) or error code. A getter comm_pair (I wrote the tests for this device with expected_protocol 😁) looks like `("RMP", "RMP: 10.5")`, a setter like `("SWA 35.658", "SWA: 35.658")` (echoing the command), and an error like `("SPS 5", "ERR: Out of Range")`. I hope these examples help understanding the communication.

If you added the read to the write, you would get a problem with values, which invokes write (which also reads) and reading afterwards again. If instead the setter invoked ask, everything would work out fine.

I'd just call it `read_set_reply`, because the variable does not cause error checking, just to read the reply. Reading should always check errors (if defined in the protocol).

**Reply by LongnoseRob** [reply]
From my understanding of this part of the discssion, I see we have 2/3 cases to differentiate here:

- Case A: SCPI complient instrument (like my R&S SFM).
    I am using the check_set_errors in [this property](https://github.com/pymeasure/pymeasure/blob/master/pymeasure/instruments/rohdeschwarz/sfm.py#L235-L250), because this command could cause an error status, if in this example a not definied system (a.k.a HW configuration) is called up.
   Here we could also use a regular call to the check the error status in the prgam flow or 
    _if supported  by instrument and pyVisa backend_ use SRQ evemt driven error status checking & handling.

- Case B: non-SCPI complient instrument, which does not echo back the communcation, but has error status read back possibility

- Case C: non-SCPI compliant instrument, which echos back the communication.
   There might be not be a dedicated error status  command/"register" and every reply from the instrument needs  be analyzed for error communciation from the instrument.

For cases A& B the `read` stays the `read` 
For case C the `read` would be need to be reworked to become `query + error detection and handling`

Now the interesting question how many of the currently supported devices fall into class A, B, or C?
And would this distrution make sense to create a dedicated instrument class with changed behavior?

**Reply by BenediktBurger** [reply]
> Now the interesting question how many of the currently supported devices fall into class A, B, or C?

Right now a handful of devices of the codebase enter category C. I use a few more devices in the lab whose drivers I plan to add.

> For case C the `read` would be need to be reworked to become `query + error detection and handling`

You're right, `write` could do the work of ask (write and read), while `ask` itself just calls `write` (and not `read`!) and `read` itself does nothing (raising an error?). I prefer, however, to keep write and read doing what their name implies.

Adding a parameter (read_set_reply or check_set_ack or whatever), which in turn switches in `Instrument.control`'s `fset` between calling `ask` and `write` is a small change and does solve the problem in a neat and simple way, without the need for another Instrument class or such.
